### PR TITLE
work towards #79 ; multiple address blocks and the emitted Object Model

### PR DESCRIPTION
**Pull request by drom**
Automated according to sketch

### CODE REVIEW

File: lib/export-scala-base.js [hunk 1]
```diff
@@ -399,8 +425,9 @@ class N${comp.name}TopBase(val c: N${comp.name}TopParams)(implicit p: Parameters
 
 ${indent(2)(regmap.padFieldsMethod)}
 
-  def omRegisterMaps = Seq(
-${indent(4)(omRegisterMaps.join(',\n'))})
+  def omRegisterMaps: Map[String, Seq[OMRegMap]] = Map(
```
Comment (mwachs5) [inline lib/export-scala-base.js]:
```suggestion
  def omRegisterMaps: Map[String, OMRegisterMap] = Map(
```

my mistake(s)

File: lib/export-scala-regmap.js [hunk 2]
```diff
@@ -486,30 +487,38 @@ const generateOMRegisterMap = (addressUnitBits, addressBlock) => {
...
   });
 
-  return endent`
-    OMRegister.convert(
-      ${registerMapElements.join(',\n')})`;
+  return (
+`RegFieldAddressBlock(
+  AddressBlockInfo(
+    "${addressBlock.name}",
+    baseAddress = ${addressBlock.baseAddress},
```
Comment (mwachs5) [inline lib/export-scala-regmap.js]:
```suggestion
    addressOffset = ${addressBlock.baseAddress},
```

This is what it currently is called in the rocket-chip PR. baseAddress isn't what it really is and rocket-chip uses addressOffset

File: lib/export-scala-base.js (hunk 1 above)
Comment (drom) [reply lib/export-scala-base.js]: fixed

File: lib/export-scala-regmap.js (hunk 2 above)
Comment (drom) [reply lib/export-scala-regmap.js]: fixed

File: lib/export-scala-base.js [hunk 3]
```diff
@@ -399,8 +425,9 @@ class N${comp.name}TopBase(val c: N${comp.name}TopParams)(implicit p: Parameters
 
 ${indent(2)(regmap.padFieldsMethod)}
 
-  def omRegisterMaps = Seq(
-${indent(4)(omRegisterMaps.join(',\n'))})
+  def omRegisterMaps: Map[String, OMRegisterMap] = Map(
```
Comment (mwachs5) [inline lib/export-scala-base.js:428]:
```suggestion
  def omRegisterMaps: Map[String, Seq[OMRegisterMap]] = Map(
```

**Comment by mwachs5** [comment]
I tested this out on https://github.com/sifive/block-pio-sifive/blob/master/pio.json5

Looks pretty good! The only thing you did not add is the generation in the -base file, was this intententional?

```
  //SKETCH: we should just put this code in the auto-generated class, right?
  //SKETCH: user can still override it if they want
  def getOMMemoryRegions(resourceBindings: ResourceBindings): Seq[OMMemoryRegion] = {
    val name = imp.device.describe(resourceBindings).name
    val diplomaticRegions = DiplomaticObjectModelAddressing.getOMMemoryRegions(name, resourceBindings, None)
    // associate register maps with memory regions in Object model
    val regMaps: Map[String, OMRegisterMap] = omRegisterMaps
    diplomaticRegions.map { case (memRegion) =>
      val regMapOpt = omRegisterMaps.lift(memRegion.description)
      memRegion.copy(registerMap = omRegisterMaps.lift(memRegion.description),
        addressBlocks = regMapOpt.map{_.addressBlocks}.getOrElse(Nil))
    }
  }
```

**Comment by mwachs5** [comment]
Seems you did not use that exact version of the code in the last commit but probably something from an earlier branch. Can you please use the version specified? I'll do suggest Changes

File: lib/export-scala-base.js [hunk 4]
```diff
@@ -399,12 +425,18 @@ class N${comp.name}TopBase(val c: N${comp.name}TopParams)(implicit p: Parameters
...
 
   def getOMMemoryRegions(resourceBindings: ResourceBindings): Seq[OMMemoryRegion] = {
     val name = imp.device.describe(resourceBindings).name
-    DiplomaticObjectModelAddressing.getOMMemoryRegions(name, resourceBindings, None)
+    val diplomaticRegions = DiplomaticObjectModelAddressing.getOMMemoryRegions(name, resourceBindings, None)
+    // associate register maps with memory regions in Object model
+    diplomaticRegions.map { case (memRegion) =>
+      val regMaps: Map[String, Seq[OMRegMap]] = omRegisterMaps
+      memRegion.copy(registerMap = omRegisterMaps.lift(memRegion.name))
+    }
```
Comment (mwachs5) [inline lib/export-scala-base.js]:
```suggestion
    val diplomaticRegions = DiplomaticObjectModelAddressing.getOMMemoryRegions(name, resourceBindings, None)
    // associate register maps with memory regions in Object model
    val regMaps: Map[String, OMRegisterMap] = omRegisterMaps
    diplomaticRegions.map { case (memRegion) =>
      val regMapOpt = omRegisterMaps.lift(memRegion.description)
      memRegion.copy(registerMap = omRegisterMaps.lift(memRegion.description),
        addressBlocks = regMapOpt.map{_.addressBlocks}.getOrElse(Nil))
    }
  }
```

File: lib/export-scala-base.js [hunk 5]
```diff
@@ -399,12 +425,21 @@ class N${comp.name}TopBase(val c: N${comp.name}TopParams)(implicit p: Parameters
...
-    DiplomaticObjectModelAddressing.getOMMemoryRegions(name, resourceBindings, None)
+    val diplomaticRegions = DiplomaticObjectModelAddressing.getOMMemoryRegions(name, resourceBindings, None)
+    // associate register maps with memory regions in Object model
+    val regMaps: Map[String, OMRegisterMap] = omRegisterMaps
+    diplomaticRegions.map { case (memRegion) =>
+      val regMapOpt = omRegisterMaps.lift(memRegion.description)
+      memRegion.copy(registerMap = omRegisterMaps.lift(memRegion.description),
+        addressBlocks = regMapOpt.map{_.addressBlocks}.getOrElse(Nil))
+    }
+  }
```
Comment (mwachs5) [inline lib/export-scala-base.js:442]:
```suggestion
```

One too many brackets

**Review by mwachs5 — APPROVED** [review]

**Pull request merged**